usr.sbin/bhyve/migration.c

```diff
@@ -0,0 +1,1259 @@
+
+			con_socket = accept(s, (struct sockaddr *)&client_sa, &client_len);
+
+			if (con_socket < 0) {
+				fprintf(stderr, "%s: Could not accept connection\r\n", __func__);
+				error = -1;
+				goto done_close_s;
+			}
+			*socket_fd = s;
+			*connection_socket_fd = con_socket;
+			break;
+		default:
+			fprintf(stderr, "%s: unknown operation request\r\n",
+				__func__);
+			error = -1;
+			goto done;
+	}
+
+	error = 0;
+	goto done;
+
+done_close_s:
+	close(s);
+done:
+	return (error);
+}
+
+int
+vm_send_migrate_req(struct vmctx *ctx, struct migrate_req req)
+{
+	int s;
+	int rc, error;
+	size_t migration_completed;
+
+
+	rc = migrate_connections(req, &s, NULL, MIGRATION_SEND_REQ);
+
+	if (rc < 0) {
+		fprintf(stderr, "%s: Could not create connection\r\n", __func__);
+		return (-1);
+	}
+
+	/* Send thw system requirements */
+	rc = migration_send_specs(s);
+
+	if (rc < 0) {
+		fprintf(stderr, "%s: Error while checking system requirements\r\n",
+			__func__);
+		error = rc;
+		goto done;
+	}
+
+	vm_vcpu_pause(ctx);
+
+	rc = vm_pause_user_devs(ctx);
+	if (rc != 0) {
+		fprintf(stderr, "Could not pause devices\r\n");
+		error = rc;
+		goto unlock_vm_and_exit;
+	}
+
+	rc = migrate_send_memory(ctx, s);
+	if (rc != 0) {
+		fprintf(stderr,
+			"%s: Could not send memory to destination\r\n",
+			__func__);
+		error = rc;
+		goto unlock_vm_and_exit;
+	}
+
+	/* Send the state of kernel related devices */
+	rc =  migrate_kern_data(ctx, s, MIGRATION_SEND_REQ);
+	if (rc != 0) {
+		fprintf(stderr,
+			"%s: Could not send kern data to destination\r\n",
+			__func__);
+		error = rc;
+		goto unlock_vm_and_exit;
+	}
+
+	/* Send the state of the userspace/emulated devices */
+	rc =  migrate_devs(ctx, s, MIGRATION_SEND_REQ);
+	if (rc < 0) {
+		fprintf(stderr,
+			"%s: Could not send pci devs to destination\r\n",
+			__func__);
+		error = rc;
+		goto unlock_vm_and_exit;
+	}
+
+	/* Wait for migration completed */
+	rc = migration_recv_data_from_remote(s, &migration_completed,
+					sizeof(migration_completed));
+	if ((rc < 0) || (migration_completed != MIGRATION_SPECS_OK)) {
+		fprintf(stderr,
+			"%s: Could not recv migration completed remote"
+			" or received error\r\n",
+			__func__);
+		error = -1;
+		goto unlock_vm_and_exit;
+	}
+
+	/* Poweroff the vm */
+	vm_destroy(ctx);
+	exit(0);
+
+unlock_vm_and_exit:
+	vm_vcpu_resume(ctx);
+
+	rc = vm_resume_user_devs(ctx);
+	if (rc != 0)
+		fprintf(stderr, "Could not resume devices\r\n");
+done:
+	close(s);
+	return (error);
+}
+
+int
+vm_recv_migrate_req(struct vmctx *ctx, struct migrate_req req)
+{
+	int s, con_socket;
+	int rc;
+	size_t migration_completed;
+
+	rc = migrate_connections(req, &s, &con_socket, MIGRATION_RECV_REQ);
+	if (rc != 0) {
+		fprintf(stderr, "%s: Could not create connections\r\n", __func__);
+		return (-1);
+	}
+
+	rc = migration_recv_and_check_specs(con_socket);
+	if (rc < 0) {
+		fprintf(stderr, "%s: Error while checking specs\r\n", __func__);
+		close(con_socket);
+		close(s);
```

Comment:
Cleanup could be done at the end of the function. All fail conditions below appear similar.